text/0000-spec-requirements.md

```diff
@@ -0,0 +1,53 @@
+# Spec Requirements
+
+## Test writing
+
+### Tests are written as framework-agnostic
+* Should support unit and end-to-end tests
+* JavaScript executed on the DOM is separate from the assertions
+* Tests accept `window` and optional DOM fragment
+    * DOM and window creation should be separate from the tests
+    * Tests should not need to know what framework the HTML was created from
+    * `window` should be used to derive `document`
+    * `window` is required within the tests to perform certain functionality in the browser (e.g. input/label connection)
+* individual tests or test group setup is dictated by the runner (unit setup vs integration setup)
+    * allow the tests to receive a setup (e.g. before/beforeAll)
+
+### Tests are written to be portable
+* Assertions do not have any dependencies outside of assertion definition (e.g. If Jest Assertions are used, the the Jest test framework does not also have to be used)
```

Comment:
and, just for clarity because the examples above don't take into account all the requirements in this RFC, it would really be like this:

## Both would retrieve the `header` in a separate function which would be

```js
/**
 * Queries DOM and returns the header's attributes
 **/
function queriesDomForThisTest () {
  const accordion = someFrameworkSpecificAccordionFun(context);
  
  return accordion.getElementsByClassName(selectors.accordion.header).getAttributes();
}
```

## when an error should be thrown:
```js
  const attributes = queriesDomForThisTest();
  const warnMsg =  'Aria-expanded should be true after the heading is clicked';
  expect(attributes['aria-expanded']), warnMsg).to.be.true;
```

## when only a warning is required
```js
  const attributes = queriesDomForThisTest();
  const warnMsg =  'Aria-expanded should be true after the heading is clicked';
  try {
    expect(attributes['aria-expanded']), warnMsg).to.be.true;
  }
  catch (e) {
    ValidationWarning(e) 
  } 
```

Reply:
> I think this is something which should be revisited. Trying to re-create an assertion library's functionality with if statement seems an unnecessary piece of added tech debt 

Is this something that we want to discuss as a requirement of the spec, or when implementing? This feels like it would tie into development best practices versus a spec requirement.

Reply:
I think it's two different choices, but both require writing requirements to determine the best way to write assertions. We know assertions are required. If we choose using an already-existing assertion library, then the requirements would be things like:

* allows separate message for each assertion
* allows returning varying message levels (like `warn` or `error`)

However, if it's determined that an assertion library should be written (as it won't take long for the `if` statements to start showing a pattern wherein we would then start to write a set of common functions to cover those common patterns...and tests to make sure those functions work correctly), then the requirements would grow considerably:

* allows separate message for each assertion
* allows returning varying message levels (like `warn` or `error`)
* can recognize a string
* can recognize a object
* can recognize a array
* can compare two strings
* can compare two objects
* can compare two arrays
* can determine a string is within an array of strings
* can determine a string matches the value of a key in an object
* can determine a string is within an array of strings of a specific key in an object
...